components/vision.py

```diff
@@ -1,36 +1,115 @@
+import time
+
 from networktables import NetworkTables
-from typing import Tuple
+from typing import Optional
+from dataclasses import dataclass
+from magicbot import feedback
+
+
+@dataclass
+class VisionData:
+    #: The distance to the target in metres.
+    distance: float
+
+    #: The angle to the target in radians.
+    angle: float
+
+    #: An arbitrary timestamp, in seconds,
+    #: for when the vision system last obtained data.
+    timestamp: float
+
+    __slots__ = ("distance", "angle", "timestamp")
 
 
 class Vision:
+    @property
```

Comment:
Just like the Pi has a separate Connection class that handles all the communication-related stuff, this should also delegate pinging and ponging and sending and measuring latency to a separate class that it instantiates and delegates to.